src/Elliptic/BoundaryConditions/AnalyticSolution.hpp

```diff
@@ -0,0 +1,234 @@
+ */
+template <typename System, size_t Dim, typename... FieldTags,
+          typename... FluxTags, typename Registrars>
+class AnalyticSolution<System, Dim, tmpl::list<FieldTags...>,
+                       tmpl::list<FluxTags...>, Registrars>
+    : public BoundaryCondition<Dim, Registrars> {
+ private:
+  using Base = BoundaryCondition<Dim, Registrars>;
+
+ public:
+  using options =
+      tmpl::list<elliptic::OptionTags::BoundaryConditionType<FieldTags>...>;
+  static constexpr Options::String help =
+      "Boundary conditions from the analytic solution";
+
+  AnalyticSolution() = default;
+  AnalyticSolution(const AnalyticSolution&) noexcept = default;
+  AnalyticSolution& operator=(const AnalyticSolution&) noexcept = default;
+  AnalyticSolution(AnalyticSolution&&) noexcept = default;
+  AnalyticSolution& operator=(AnalyticSolution&&) noexcept = default;
+  ~AnalyticSolution() noexcept = default;
+
+  /// \cond
+  explicit AnalyticSolution(CkMigrateMessage* m) noexcept : Base(m) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(AnalyticSolution);
+  /// \endcond
+
+  /// Select which `elliptic::BoundaryConditionType` to apply for each field
+  explicit AnalyticSolution(
+      const typename elliptic::OptionTags::BoundaryConditionType<
+          FieldTags>::type... boundary_condition_types) noexcept
+      : boundary_condition_types_{boundary_condition_types...} {}
+
+  std::unique_ptr<domain::BoundaryConditions::BoundaryCondition> get_clone()
+      const noexcept override {
+    return std::make_unique<AnalyticSolution>(*this);
+  }
+
+  const auto& boundary_condition_types() const noexcept {
+    return boundary_condition_types_;
+  }
+
+  using argument_tags =
+      tmpl::list<::Tags::AnalyticSolutionsBase, domain::Tags::Mesh<Dim>,
+                 domain::Tags::Direction<Dim>,
+                 ::Tags::Normalized<domain::Tags::UnnormalizedFaceNormal<
+                     Dim, Frame::Inertial>>>;
+  using volume_tags =
+      tmpl::list<::Tags::AnalyticSolutionsBase, domain::Tags::Mesh<Dim>>;
+
+  template <typename OptionalAnalyticSolutions>
+  void apply(const OptionalAnalyticSolutions& optional_analytic_solutions,
+             const Mesh<Dim>& volume_mesh, const Direction<Dim>& direction,
+             const tnsr::i<DataVector, Dim>& face_normal,
+             const gsl::not_null<typename FieldTags::type*>... fields,
+             const gsl::not_null<typename FieldTags::type*>... n_dot_fluxes)
+      const noexcept {
+    const auto& analytic_solutions =
+        [&optional_analytic_solutions]() noexcept -> decltype(auto) {
+      if constexpr (tt::is_a_v<std::optional, OptionalAnalyticSolutions>) {
+        if (not optional_analytic_solutions.has_value()) {
+          ERROR(
+              "The background does not provide an analytic solution. Select an "
+              "analytic solution as background if you want to use it to impose "
+              "boundary conditions.");
```

Comment:
should this say `analytic boundary conditions`? Currently it reads like you cannot impose boundary conditions at all if the background doesn't have an analytic solution. I'm also confused why this is the first time `background` shows up. It's not clear what's special about the background here, vs the solution variables. My understanding was you impose boundary conditions on the solution variables. This seems like it's meant to be a user-facing error, if that is the case this needs to be a lot clearer and should maybe use `ERROR_NO_TRACE` instead